src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.GeneratorDriverCreationCache.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed partial class SolutionCompilationState
+{
+    internal sealed class GeneratorDriverCreationCache
+    {
+        // HACK: we probably shouldn't have this be a static instance, but this is easier than threading it through from the Workspace layer.
+        internal static readonly GeneratorDriverCreationCache Instance = new();
+
+        private ImmutableDictionary<ProjectId, TaskCompletionSource<GeneratorDriver>> _driverCache = ImmutableDictionary<ProjectId, TaskCompletionSource<GeneratorDriver>>.Empty;
```

Comment:
Haven't looked at the whole pr, but we rarely cache TCS.  We cache AsyncLazy instead.